### DISCUSSION TITLE
Does R code completion work in VS Code?

### DISCUSSION THREAD

**Discussion by joakimwallmark** [quarto-dev/quarto-cli#4269] [Q&A]
I'm running quarto with vscode and I can't seem to get any code completion for R. It works for python code.

I'm on windows 11 and I have installed the quarto and R vscode extensions. I do have code completion from the R extension in R and rmd files, just not in qmd. However, qmd documents with r code do knit and I can also execute the code within code cells. Do I need to use RStudio for R programming if I want code completion or do I have some other issue?

**Comment by shrikantad** [comment]
I am having the same issue. I can run the R code in terminal from VSCode. However, the .qmd docs don't have auto-completion. Tab completion and linting would be extremely helpful. Is there a VSCode extension for this?

**Comment by cderv** [comment]
Quarto Extension uses other language extenstion for code cells https://quarto.org/docs/tools/vscode.html#code-cells

So using the R extension (https://marketplace.visualstudio.com/items?itemName=REditorSupport.r) should be enough. 

I am also on Windows 11 and have both extension, and I correctly get the completion either while typing in chunk, or doing CTRL+Space to trigger it.

**Comment by karshy72** [comment]
I am having a similar issue. The base R functions do get suggested so its not entirely broken. Objects that exist in the name space are also coming up as suggestions in the autocomplete. However, functions that exist inside of packages which have been loaded do not appear. So for example, after loading tidyverse with library(tidyverse) typing "mut" does not suggest an autocomplete to mutate. However, if I were to type "dplyr::mut" then mutate would be suggested as an autocomplete.

**Reply by mcanouil** [reply]
As highlighted before, R auto-complete feature comes from <https://marketplace.visualstudio.com/items?itemName=REditorSupport.r>.
Unless this behaviour only appears in Quarto document while using both Quarto and R VSCode extensions, not much can be done here.
Also, providing a step by step approach to reproduce would help, as would providing information about what R/Quarto/VSCode/VSCode extensions/OS/etc. version you are using.

**Reply by karshy72** [reply]
this only happens in quarto but not in rmarkdown documents or in raw r files. 

Details of the system and versions below. 

VS Code v1.77.0
Quarto v1.79.0
R 4.2.0
VS Code R extension v2.7.2 

On M1 Mac Book Pro running Ventura 13.0.1

Reproducible example of when this happens. I have attached a file with .md as the extension as it wont let me upload a qmd file but if you switch it to qmd before opening should reproduce what I am talking about.

[quarto_R_autocomplete_example.md](https://github.com/quarto-dev/quarto-cli/files/11140384/quarto_R_autocomplete_example.md)

**Reply by mcanouil** [reply]
Ok, I can reproduce this.
Switching file type as rmarkdown instead of quarto in VSCode to use vscode-r directly allows completion, so there is something going on.

https://user-images.githubusercontent.com/8896044/232908760-8c673bc8-27db-4523-bae9-b5bec01b696e.mov

@jjallaire Do you have an idea of what could cause this in Quarto VSCode extension?

**Reply by jjallaire** [reply]
I sadly don't have any idea what could be going on here! With Quarto we write a temp .R file and call for completions within that file. This should in principle be the same but clearly something about the calling context is resulting in different completions. It may be worth following up w/ the folks developing the R extension to see if they have any theories about what is going on and/or what we might be able to do differently. 

One thought: what if the R extension simply supported completions in `.qmd` the same as they do for `.Rmd`? If that were the case then we could stop writing our temp file and just rely on that behavior?

**Reply by mcanouil** [reply]
> One thought: what if the R extension simply supported completions in .qmd the same as they do for .Rmd? If that were the case then we could stop writing our temp file and just rely on that behavior?

I think this might be the best solution, _i.e._, I was wondering why some times in Quarto documents I had an "intellisense.R" file created and opened in VSCode, from what you are saying, it seems to be the temporary file.

**Reply by cderv** [reply]
> One thought: what if the R extension simply supported completions in .qmd the same as they do for .Rmd? If that were the case then we could stop writing our temp file and just rely on that behavior?

I believe it was on the table, then off due to Quarto related work, but it would be worth reopening or commenting in the related issue. Code completion for R and Rmd in VSCODE is done by the **languageserver** package
https://github.com/REditorSupport/languageserver/issues/575

We could also try contribute this.
